src/schema/users.ts

```diff
@@ -850,14 +850,61 @@ export const resolvers: IResolvers<any, Context> = {
         .orderBy('date')
         .getRawMany();
     },
-    userStreak: (_, __, ctx: Context, info): Promise<GQLUserStreak> =>
-      graphorm.queryOneOrFail<GQLUserStreak>(ctx, info, (builder) => ({
-        queryBuilder: builder.queryBuilder.where(
-          `"${builder.alias}"."userId" = :id`,
-          { id: ctx.userId },
-        ),
-        ...builder,
-      })),
+    userStreak: async (_, __, ctx: Context, info): Promise<GQLUserStreak> => {
+      const streak = await graphorm.queryOneOrFail<GQLUserStreak>(
+        ctx,
+        info,
+        (builder) => ({
+          queryBuilder: builder.queryBuilder.where(
+            `"${builder.alias}"."userId" = :id`,
+            { id: ctx.userId },
+          ),
+          ...builder,
+        }),
+      );
+
+      const lastView = await getStreakLastView(ctx.con, streak);
+
+      if (!lastView) {
+        return streak;
+      }
+
+      const today = new Date();
+      const dayToday = today.getDay();
+      const difference = differenceInDays(today, lastView);
+
+      if (difference === 0) {
+        return streak;
+      }
+
+      // Even though it is the weekend, we should still clear the streak for when the user's last read was Thursday
+      // Due to the fact that when Monday comes, we will clear it anyway when we notice the gap in Friday
+      if (WEEKENDS.includes(dayToday)) {
```

Comment:
> You could then argue, that we could also do everything in a simple if/else condition as well. Much quicker to read and I think readability is actually improved, since you don't need to read so many lines:
```
if ((today === Saturday && difference > 1) ||
     (today === Sunday && difference > 2) || 
     (today === Monday && difference > 3) ||
     difference > 1) {
     return clearThenGetUserStreak(ctx, info, ctx.userId);
}

return streak; 
```
Yeah, this makes sense and much simpler. Let's apply it.

Reply:
> This condition is unnecessary. I would flatten the conditions to:
> 
> ```js
> const clearTheStreak = () => clearThenGetUserStreak(ctx, info, ctx.userId)
> 
> if (today === Saturday && difference > 1) {
>   return clearTheStreak();
> }
> if (today === Sunday && difference > 2) {
>   return clearTheStreak();
> }
> if (today === Monday && difference > 3) {
>   return clearTheStreak();
> }
> if (difference > 1) {
>   return clearTheStreak();
> }
> 
> return streak;
> ```
> 
> You could then argue, that we could also do everything in a simple if/else condition as well. Much quicker to read and I think readability is actually improved, since you don't need to read so many lines:
> 
> ```js
> if ((today === Saturday && difference > 1) ||
>      (today === Sunday && difference > 2) || 
>      (today === Monday && difference > 3) ||
>      difference > 1) {
>      return clearThenGetUserStreak(ctx, info, ctx.userId);
> }
> 
> return streak;    
> ```

This looks great but it won't work when it is a Weekend the user's last read was Friday where it should not reset.

Reply:
Good catch. But it will work, with a slight modification:
```javascript
if ((dayToday === Sunday && difference > 2) || 
     (dayToday === Monday && difference > 3) ||
     (dayToday > Monday && dayToday <= Saturday && difference > 1)) {
     return clearThenGetUserStreak(ctx, info, ctx.userId);
}

return streak; 
```

So now, if it's Sunday, the difference must be more than 2 days to clear.
If it's Monday, the difference must be more than 3 days to clear.
If it's any other day, it can be more than 1 day to clear.

I think we are covered by this. We can split it into multiple conditions as well, if you find it more readable, but I like this concise version more, me being me :)

Reply:
Probably a matter of preference, but I think your suggestion is much clearer, and I'd prefer it more. I applied the suggestion then just modified some of the magic numbers just to make it easier to read.

Reply:
The conditions for clearing are now 👨‍🍳 💋 , love it ❤️

We still need to fix the user timezone issue on the date comparison though. https://github.com/dailydotdev/daily-api/pull/1658#discussion_r1471548004